module/zfs/metaslab.c

```diff
@@ -1659,9 +1654,92 @@ metaslab_block_picker(range_tree_t *rt, uint64_t *cursor, uint64_t size,
 	*cursor = 0;
 	return (-1ULL);
 }
-#endif /* WITH_DF/CF_BLOCK_ALLOCATOR */
 
-#if defined(WITH_DF_BLOCK_ALLOCATOR)
+typedef enum spa_allocator {
+	SPA_DF_ALLOCATOR = 0,	/* Dynamic Fit (df) block allocator */
+	SPA_CF_ALLOCATOR,	/* Cursor fit block allocator */
+	SPA_NDF_ALLOCATOR,	/* New dynamic fit allocator */
+	SPA_NUM_ALLOCATORS
+} spa_allocator_t;
+
+static uint64_t metaslab_df_alloc(metaslab_t *msp, uint64_t size);
+static uint64_t metaslab_cf_alloc(metaslab_t *msp, uint64_t size);
+static uint64_t metaslab_ndf_alloc(metaslab_t *msp, uint64_t size);
+metaslab_ops_t *metaslab_allocator(spa_t *spa);
+
+static metaslab_ops_t metaslab_allocators[SPA_NUM_ALLOCATORS] = {
+	{ "dynamic", metaslab_df_alloc },
+	{ "cursor", metaslab_cf_alloc },
+	{ "new-dynamic", metaslab_ndf_alloc },
+};
+
+static int
+spa_find_allocator_byname(const char *val)
+{
+	for (int a = 0; a < SPA_NUM_ALLOCATORS; a++) {
+
+		/* todo: for removal when ndf is working */
+		if (a == SPA_NDF_ALLOCATOR) continue;
+
+		if (strcmp(val, metaslab_allocators[a].msop_name) == 0)
+			return (a);
+	}
+	return (-1);
+}
+
+void
+spa_set_allocator(spa_t *spa, const char *allocator)
+{
+	int a = spa_find_allocator_byname(allocator);
+	if (a < 0) a = SPA_DF_ALLOCATOR;
+	spa->spa_active_allocator = a;
+	zfs_dbgmsg("spa allocator: %s\n", metaslab_allocators[a].msop_name);
+}
+
+uint64_t
+spa_get_active_allocator(spa_t *spa)
```

Comment:
Wonder why here it is "active_allocator", while above just "allocator"?

Reply:
renamed, thanks